### PR TITLE
Parallelize run_test

### PR DESCRIPTION
**Pull request by Smit-create**
On this branch:
```
% time ./run_tests.py -j 16
./run_tests.py -j 16  14.79s user 8.29s system 560% cpu 4.120 total
```

On main:
```
% time ./run_tests.py
./run_tests.py  9.21s user 4.05s system 75% cpu 17.536 total
```

### CODE REVIEW

**Comment by certik** [comment]
That's fine. But we need to test it carefully, so I can do that tomorrow.

**Comment by czgdp1807** [comment]
Much needed. Thanks.

**Review by czgdp1807 — APPROVED** [review]
Let's wait for @certik's approval as well before merging.

**Comment by certik** [comment]
The LFortran's runner has been parallelized with `concurrent.futures`. So I ready here: https://stackoverflow.com/questions/20776189/concurrent-futures-vs-multiprocessing-in-python-3 that the two approaches are essentially equivalent.  We should strive to merge/move as much as possibly from LPython/LFortran into libasr, so that we have a common codebase sa much as possible, not diverge it.

One big problem with this PR is that it does not (easily) tell you the diff. Consider:
```
TEST: c_interop1.py
    * asr    ✓ 
    * ast    The JSON metadata differs against reference results
Reference JSON: tests/reference/ast-expr4-49316cb.json
Output JSON:    tests/output/ast-expr4-49316cb.json
Omitting 9 identical items
Differing items:
{'infile_hash': '14623118176bce0eaf98012ed91be759304ecb0abbcb97835a9daf43'} != {'infile_hash': '5cba7a5d589f54fc31463e48903d5b46604fb64e3e64ba215339047c'}
{'stdout_hash': '8ab9450e7b985d0f80e6ea3e9607235ca660802ac7a179741c59b14c'} != {'stdout_hash': '0021b494511ff634740b74baed3fcad55b0559c874433a63756f257d'}
Diff against: tests/reference/ast-expr4-49316cb.stdout
1c1
< (Module [(FunctionDef test_del ([] [] [] [] [] [] []) [(AnnAssign (Name a Store) (Name i32 Load) () 1) (AnnAssign (Name b Store) (Name i32 Load) () 1) (Assign [(Name a Store)] (ConstantInt 4 ()) ()) (Assign [(Name b Store)] (ConstantInt 20 ()) ()) (Delete [(Name a Del) (Name b Del)])] [] () ())] [])
---
> (Module [(FunctionDef test_del ([] [] [] [] [] [] []) [(AnnAssign (Name a Store) (Name i32 Load) () 1) (AnnAssign (Name b Store) (Name i32 Load) () 1) (Assign [(Name a Store)] (ConstantInt 4 ()) ()) (Assign [(Name b Store)] (ConstantInt 20 ()) ()) (Delete [(Name a Del) (Name c Del)])] [] () ())] [])
    * ast    ✓ 
TEST: ../integration_tests/test_complex.py
    * ast    ✓ 
    * llvm   ✓ 



[pages of output]


multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/Users/ondrej/mambaforge/envs/lp/lib/python3.10/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/Users/ondrej/mambaforge/envs/lp/lib/python3.10/multiprocessing/pool.py", line 48, in mapstar
    return list(map(*args))
  File "/Users/ondrej/repos/lpython/run_tests.py", line 51, in __call__
    run_test("ast", "lpython --show-ast --no-color {infile} -o {outfile}",
  File "/Users/ondrej/repos/lpython/compiler_tester/tester.py", line 273, in run_test
    raise RunException("The reference result differs")
compiler_tester.tester.RunException: The reference result differs
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/ondrej/repos/lpython/./run_tests.py", line 136, in <module>
    main()
  File "/Users/ondrej/repos/lpython/./run_tests.py", line 122, in main
    pool.map(tester, d["test"])
  File "/Users/ondrej/mambaforge/envs/lp/lib/python3.10/multiprocessing/pool.py", line 364, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "/Users/ondrej/mambaforge/envs/lp/lib/python3.10/multiprocessing/pool.py", line 771, in get
    raise self._value
compiler_tester.tester.RunException: The reference result differs
```



Consider LFortran's parallel run:
```
Traceback (most recent call last):
  File "/Users/ondrej/repos/lfortran/lfortran/./run_tests.py", line 281, in <module>
    main()
  File "/Users/ondrej/repos/lfortran/lfortran/./run_tests.py", line 267, in main
    for f in futures:
  File "/Users/ondrej/mambaforge/envs/lf/lib/python3.10/concurrent/futures/_base.py", line 609, in result_iterator
    yield fs.pop().result()
  File "/Users/ondrej/mambaforge/envs/lf/lib/python3.10/concurrent/futures/_base.py", line 446, in result
    return self.__get_result()
  File "/Users/ondrej/mambaforge/envs/lf/lib/python3.10/concurrent/futures/_base.py", line 391, in __get_result
    raise self._exception
  File "/Users/ondrej/mambaforge/envs/lf/lib/python3.10/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/ondrej/repos/lfortran/lfortran/./run_tests.py", line 50, in single_test
    run_test(
  File "/Users/ondrej/repos/lfortran/lfortran/compiler_tester/tester.py", line 307, in run_test
    raise RunException(
compiler_tester.tester.RunException: Testing with reference output failed.
subroutine1.f90 * tokens
The JSON metadata differs against reference results
Reference JSON: tests/reference/tokens-subroutine1-c249ad5.json
Output JSON:    tests/output/tokens-subroutine1-c249ad5.json
Omitting 9 identical items
Differing items:
{'infile_hash': '57df3f8d604e3cd0919c00fcf6836b70520ca803b4682b7b00fbe644'} != {'infile_hash': 'e209c22328b423e75d6801f3c0ef19bc120936196435dff2cb19df31'}
{'stdout_hash': '6f98c92c044d6ba2ec5b53a0ad43e11243aa7629fba9c0bc8974f422'} != {'stdout_hash': 'e48e3d92e7b293dcc3b2c28a10054c5191a8c2bda04d3a63a8c1b50e'}
Diff against: tests/reference/tokens-subroutine1-c249ad5.stdout
12c12
< (TOKEN "identifier" x)
---
> (TOKEN "identifier" yx)

```



So instead of implementing the same feature in a slightly different way (different command line options, etc.), why not take what is in LFortran (which works and looks reasonable), and also do the work and merge more things into the libasr directory? We effectively have the `compiler_tester` directory, which is shared (I think), so let's move that to libasr, and move the parallel stuff in there as well, so that the only thing that is different in LFortran/LPython is just the actual tests definitions?

That way if we add another frontend to LCompilers, all the infrastructure gets reused.

**Comment by czgdp1807** [comment]
I agree with @certik regarding reusing parallelization of tests from LFortran.

**Comment by Smit-create** [comment]
> We effectively have the compiler_tester directory, which is shared (I think), so let's move that to libasr, and move the parallel stuff in there as well, so that the only thing that is different in LFortran/LPython is just the actual tests definitions?

Yes, I agree to that.

**Comment by Smit-create** [comment]
On this diff:
```
% time python compiler_tester/run_tests.py
python compiler_tester/run_tests.py  13.43s user 6.73s system 573% cpu 3.518 total
```
and,
```
% time ./run_tests.py
./run_tests.py  9.02s user 3.86s system 81% cpu 15.827 total
```

**Comment by certik** [comment]
Thanks!

I will test it later today. We also need to update LFortran with the same, to share exactly the same code.

File: compiler_tester/run_tests.py
Comment (certik) [inline compiler_tester/run_tests.py:173]: I am worried that once we have 10 frontends, we'll have to maintain the union of these.

I think the `compiler_tester` should contain code that is used by all frontends, the "common" code. The above does not look like common code.

If I have time, I'll try to help with this to get this done.

**Comment by certik** [comment]
I fixed this in https://github.com/lcompilers/lpython/pull/909. @Smit-create go ahead and review it please.

**Comment by certik** [comment]
I think this is now fixed. If you have ideas for improvements, or speed improvements, definitely submit more PRs. The LFortran/LPython specific test runner is now pretty simple, most common code is now shared in `libasr`.

**Pull request closed without merging**
